crates/bevy_pbr/src/material_bind_groups.rs

```diff
@@ -155,634 +276,1152 @@ impl From<u32> for MaterialBindGroupIndex {
+            self.allocated_resource_count += 1;
+        }
+
+        allocated_resource_slots
     }
 
-    /// Marks the given slot as free.
-    fn free(&mut self, slot: MaterialBindGroupSlot) {
-        self.unprepared_bind_groups[slot.0 as usize] = None;
-        self.used_slot_bitmap &= !(1 << slot.0);
+    /// Removes the material allocated in the given slot, with the given
+    /// descriptor, from this slab.
+    fn free(&mut self, slot: MaterialBindGroupSlot, bindless_descriptor: &BindlessDescriptor) {
+        // Loop through each binding.
+        for (bindless_index, (bindless_resource_type, &bindless_binding)) in bindless_descriptor
+            .resources
+            .iter()
+            .zip(self.bindless_index_table.get(slot, bindless_descriptor))
+            .enumerate()
+        {
+            let bindless_index = BindlessIndex::from(bindless_index as u32);
+
+            // Free the binding.
+            let resource_freed = match *bindless_resource_type {
+                BindlessResourceType::None => false,
+                BindlessResourceType::Buffer => self
+                    .buffers
+                    .get_mut(&bindless_index)
+                    .expect("Buffer should exist with that bindless index")
+                    .remove(bindless_binding),
+                BindlessResourceType::SamplerFiltering
+                | BindlessResourceType::SamplerNonFiltering
+                | BindlessResourceType::SamplerComparison => self
+                    .samplers
+                    .get_mut(bindless_resource_type)
+                    .expect("Sampler array should exist")
+                    .remove(bindless_binding),
+                BindlessResourceType::Texture1d
+                | BindlessResourceType::Texture2d
+                | BindlessResourceType::Texture2dArray
+                | BindlessResourceType::Texture3d
+                | BindlessResourceType::TextureCube
+                | BindlessResourceType::TextureCubeArray => self
+                    .textures
+                    .get_mut(bindless_resource_type)
+                    .expect("Texture array should exist")
+                    .remove(bindless_binding),
+            };
+
+            // If the slot is now free, decrement the allocated resource
+            // count.
+            if resource_freed {
+                self.allocated_resource_count -= 1;
+            }
+        }
 
-        // Invalidate the cached bind group so that we rebuild it again.
+        // Clear out the extra data.
+        self.extra_data[slot.0 as usize] = None;
+
+        // Invalidate the cached bind group.
         self.bind_group = None;
-    }
 
-    /// Returns true if all the slots are full or false if at least one slot in
-    /// this bind group is free.
-    fn is_full(&self) -> bool {
-        self.used_slot_bitmap == (1 << (self.unprepared_bind_groups.len() as u32)) - 1
+        // Release the slot ID.
+        self.free_slots.push(slot);
```

Comment:
If a bad descriptor is passed in somehow and no resources are found in the table, this code to push the slot and decrement the allocation count will still run. I *think* this is still correct, but wanted to highlight for review in case this is a real edge case.

Reply:
The use of ref counting means that double frees are going to cause havoc in general unfortunately. We just have to be careful.